### ISSUE TITLE
kernel BUG at /root/tempesta/tempesta_fw/cache.c:620

### ISSUE THREAD

**Issue by keshonok** [natsys/tempesta#482]
This occurs on the current master at 3c934b8.

The configuration is as follows:
```
   server 106.185.29.4:80 ;
   cache 1;

   sticky name=__cookie__ enforce;
```

Tempesta is tested with a simple single `wget -d -S http://127.0.0.1`. First wget completes successfully, but the second one leads to a crash.

I also tried to reset the DB by removing it before starting Tempesta. That didn't change anything.

```
[  307.959034] kernel BUG at /root/tempesta/tempesta_fw/cache.c:620!
[  307.959051] invalid opcode: 0000 [#1] SMP
[  307.959065] Modules linked in: tfw_sched_rr(OE) tfw_sched_http(OE) tfw_sched_hash(OE) tempesta_fw(OE) tempesta_db(OE) mgag200 x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel syscopyarea kvm sysfillrect sysimgblt crct10dif_pclmul crc32_pclmul crc32c_intel ttm drm_kms_helper aesni_intel drm lrw gf128mul glue_helper ablk_helper cryptd iTCO_wdt iTCO_vendor_support ipmi_ssif i2c_i801 ipmi_si pcspkr wmi sg sb_edac edac_core ipmi_msghandler mei_me mei ioatdma lpc_ich mfd_core shpchp ip_tables ext4 mbcache jbd2 raid1 sd_mod isci igb libsas ahci i2c_algo_bit libahci i2c_core scsi_transport_sas libata dca serio_raw ptp pps_core
[  307.959293] CPU: 1 PID: 0 Comm: swapper/1 Tainted: G           OE   4.1.12+ #1
[  307.959324] Hardware name: Supermicro X9SRE/X9SRE-3F/X9SRi/X9SRi-3F/X9SRE/X9SRE-3F/X9SRi/X9SRi-3F, BIOS 3.2 01/16/2015
[  307.959351] task: ffff88083b90a910 ti: ffff88083b918000 task.ti: ffff88083b918000
[  307.959370] RIP: 0010:[<ffffffffa0475e6c>]  [<ffffffffa0475e6c>] tfw_cache_write_field+0xec/0x120 [tempesta_fw]
[  307.959401] RSP: 0018:ffff88087fc43d48  EFLAGS: 00010246
[  307.959415] RAX: ffff88087fc43e68 RBX: ffff88083770c578 RCX: 0000000000000000
[  307.959433] RDX: 00000000000000f3 RSI: ffff88083770c020 RDI: ffff88087fc43e80
[  307.959451] RBP: ffff88087fc43dc8 R08: ffff880839324cf0 R09: 0000000000000000
[  307.959469] R10: ffff88083251b092 R11: ffff88087fc43d70 R12: 0000000000000061
[  307.959487] R13: 0000000000000061 R14: ffff88087fc43e60 R15: ffff88084e43e000
[  307.959506] FS:  0000000000000000(0000) GS:ffff88087fc40000(0000) knlGS:0000000000000000
[  307.959526] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  307.959541] CR2: 00007f0def49b000 CR3: 00000000019ba000 CR4: 00000000001406e0
[  307.959560] Stack:
[  307.959566]  ffff880833d4a480 ffff88087fc43e68 ffff88087fc43e80 ffff88083770c020
[  307.959589]  ffff88087fc43d88 ffff88084e43e010 0000000000000000 0000000000000000
[  307.960332]  0000000000000000 a0f201ea212cf2ed 0000000000000000 ffff88083770c2b8
[  307.961076] Call Trace:
[  307.961811]  <IRQ>
[  307.961820]  [<ffffffffa047730f>] tfw_wq_tasklet+0x98f/0xa90 [tempesta_fw]
[  307.963239]  [<ffffffffa047c3b0>] ? tfw_http_msg_process+0x620/0x620 [tempesta_fw]
[  307.963961]  [<ffffffffa047c3b0>] ? tfw_http_msg_process+0x620/0x620 [tempesta_fw]
[  307.964677]  [<ffffffff81088c06>] tasklet_action+0xe6/0xf0
[  307.965391]  [<ffffffff81088ea4>] __do_softirq+0xf4/0x2d0
[  307.966107]  [<ffffffff81089385>] irq_exit+0x125/0x130
[  307.966823]  [<ffffffff8104fa65>] smp_call_function_single_interrupt+0x35/0x40
[  307.967531]  [<ffffffff81690cbe>] call_function_single_interrupt+0x6e/0x80
[  307.968223]  <EOI>
[  307.968232]  [<ffffffff81520d79>] ? cpuidle_enter_state+0xa9/0x1f0
[  307.969573]  [<ffffffff81520d48>] ? cpuidle_enter_state+0x78/0x1f0
[  307.970236]  [<ffffffff81520ef7>] cpuidle_enter+0x17/0x20
[  307.970898]  [<ffffffff810ca56c>] cpu_startup_entry+0x35c/0x3f0
[  307.971567]  [<ffffffff810503e3>] start_secondary+0x173/0x1e0
[  307.972239] Code: 85 c0 75 2d 48 8b 45 b8 49 01 06 41 01 c5 49 63 c5 48 3b 43 10 74 18 45 8b 4f 08 45 85 c9 75 87 48 8b 45 88 48 c7 00 00 00 00 00 <0f> 0b 66 90 31 c0 48 8b 7d c8 65 48 33 3c 25 28 00 00 00 75 0f
[  307.973724] RIP  [<ffffffffa0475e6c>] tfw_cache_write_field+0xec/0x120 [tempesta_fw]
[  307.974437]  RSP <ffff88087fc43d48>
```

**Comment by milabs** [comment]
@sergsever As the problem was caused by the duplicate HTTP-headers and incorrect restoring from the cache, it would be nice to have a test which covers that case. Here is the real HTTP-header which caused the crash:

```
HTTP/1.1 200 OK
Server: Tengine/1.3.0
Date: Wed, 18 May 2016 21:18:38 GMT
Content-Type: text/html; charset=gbk
Transfer-Encoding: chunked
Connection: keep-alive
X-Powered-By: PHP/5.3.14
Set-Cookie: MECz_2132_saltkey=P74ZCDY8; expires=Fri, 17-Jun-2016 21:18:38 GMT; path=/; httponly
Set-Cookie: MECz_2132_lastvisit=1463602718; expires=Fri, 17-Jun-2016 21:18:38 GMT; path=/
Set-Cookie: MECz_2132_sid=CmGJJc; expires=Thu, 19-May-2016 21:18:38 GMT; path=/
```

**Comment by keshonok** [comment]
The core of this bug is fixed with 7b4cb41.

**Comment by keshonok** [comment]
The request for a test has been set aside as a separate issue #490.
